### PR TITLE
(Slightly) optimize catch and try/catch

### PR DESCRIPTION
**Pull request by bjorng**
This PR contains several improvements to catch and try/catch. The commit message for each commit gives more details. Here are the highlights:

* A catch that ignores its return value such as this code:

```
catch side_effect(),
...

```

will no longer build a stack trace.

* There are several minor optimizations, both in the compiler and run-time system, for try/catch.

* The optimization of receive that can avoid rescanning the entire message queue will now work for this code:

         Ref = make_ref(),
         try
             side_effect()
         catch _:_ ->
              ok
         end,
         receive
            %% Clauses that all match Ref.
            .
            .
            .
         end

(try/catch used to prevent the optimization.)

### CODE REVIEW

**Comment by bjorng** [comment]
@kostis @mikpe @richcarl: I had to modify `hipe_beam_to_icode` to keep the hipe compiler working, but I am not sure that my fix is always correct (the fix for hipe is in the fixup commit). I have compiled with `--enable-native-libs` and have run the test suite for `hipe` without finding any problems. I would appreciate if you could have a look at my fix.

File: erts/emulator/beam/instrs.tab [hunk 1]
```diff
@@ -888,12 +888,16 @@ catch_end(Y) {
...
+}
+
+try_case(Y) {
+    c_p->catches--;
+    make_blank($Y);
+    ASSERT(is_non_value(r(0)));
+    c_p->fvalue = NIL;
+    r(0) = x(1);
+    x(1) = x(2);
+    x(2) = x(3);
```
Comment (garazdawi) [inline erts/emulator/beam/instrs.tab]: maybe rewrite as

    try_case(Y) {
      $try_end($Y);
      ASSERT(is_non_value(r(0)));
      ....

and the same for catch_end.

**Comment by richcarl** [comment]
Maybe @happi has some vague memories of how beam-to-icode works...

File: erts/emulator/beam/instrs.tab (hunk 1 above)
Comment (bjorng) [reply erts/emulator/beam/instrs.tab]: Good idea!

**Comment by kostis** [comment]
On the surface the change in `hipe_beam_to_icode` looks OK but it has been really long back since I last looked at that code.  I am slightly concerned about the fact that this PR comes without any tests for its functionality.  If such tests are written, perhaps they can also be added to the HiPE testsuite which will hopefully give us a little more confidence that the change is OK.


Aside: I found this thread by accident. Since the beginning of summer or so, I do not get notifications from GitHub not only from repositories I watch but even when explicitly mentioned by username in the comments (as here).  I've tried to look at the settings of my account, but they look OK.  Any help/ideas about it are appreciated (please send by private mail).

**Comment by bjorng** [comment]
I did not write any new test cases before there is no new functionality, just an optimization of existing functionality. (Of course, that could still mean that new tests could be needed, but it is far from obvious what the test cases should test that is not already tested.)

catch is used in many places in the test suites. try/catch is tested extensively in trycatch_SUITE in the compiler test suites (21 test cases). There are also test of generating and catching exceptions in exception_SUITE in the emulator test suite (11 test cases). I have checked the coverage of my new code in the compiler to verify that it is indeed executed.

I have run trycatch_SUITE compiled with `native`. The test case `eclectic/1` fails, both with this PR and without. The other 20 test cases pass.

I tried to compile exception_SUITE to native code, but the compilation never finishes (same result with and without this PR).

**Comment by kostis** [comment]
> I have run trycatch_SUITE compiled with native. The test case eclectic/1 fails, both with this PR and without. 

Isn't this a problem? Does this mean there is a known bug in the compiler or is simply that the test is broken?

> I tried to compile exception_SUITE to native code, but the compilation never finishes ...

I hope to find some time to look into this.

**Comment by kostis** [comment]
> > I have run trycatch_SUITE compiled with native. The test case eclectic/1 fails, both with this PR and without.
>
> Isn't this a problem? Does this mean there is a known bug in the (native code) compiler or is it simply that the test is broken and/or shows a discrepancy with the exceptions that BEAM generates?

I've looked at this failing test case (in native code) and discovered that this test has actually changed behavior between Erlang/OTP versions.

With the help of @bjorng, commit https://github.com/erlang/otp/commit/e8d45ae14c6c3bdfcbbc7964228b004ef4f11ea6 (done by @josevalim) was identified as the one introducing the change in the information that exceptions generated by BEAM produce (and break the `eclectic/1` test case). The same change was not considered for native code, which most likely reveals something about how/which tests are run.  Despite this issue, note that such changes are _fundamentally_ problematic: they are not easy to do in _any_ native code compiler (HiPE, JiT, etc.) that compiles arithmetic to native code, and may also be a problem if one wants to have code that works similarly across different Erlang/OTP releases.

I am not sure what's the proper action here...

**Comment by josevalim** [comment]
I am also unsure what is the proper action but I can add that the rationale for my changes was to improve debugging of Erlang/Elixir codebases. The fact those operations did not include a stacktrace entry made some cases really hard to debug. Sometimes all you get is `badarith` error, without a nearby line and without the operation arguments.

I also agree the changes are problematic given we were unable to add stacktraces to all BEAM ops codes. Some ops codes had proper stacktraces before my pull request and I was able to improve others. But in some cases the optimizations done by the compiler made it impossible to rebuild the user's original call. Still, I would be sad to see those changes reverted because I believe they bring real value.

I guess at the end of the day the question we need to answer is what are the guarantees we provide for stacktraces. I would say not many, given the combination of last call optimization, stacktrace depth and the inconsistency with ops code stacktraces even before my changes.

I would generally say relying on the stacktrace entries for any reason other than debugging is a no-no, except maybe in a handful of cases such as the first entry of `error:undef`. Therefore I propose it is acceptable for native code/jit/erjang to provide less information in certain cases and, if so, we document those guarantees.

**Comment by kostis** [comment]
It's probably a bad idea to mix not just one, but two relatively unrelated to this PR discussions here, but I'll do it anyway.

> I tried to compile exception_SUITE to native code, but the compilation never finishes ...

"Never finishes" happens alright (if one uses the native code compiler in a particular way), but it's a bit inaccurate. Using my favorite way to invoke the HiPE compiler, I get a compiler crash:
<pre>
HiPE (v 3.15.3)  Error: [hipe:870]: INTERNAL ERROR
while compiling exception_SUITE:my_times/2
crash reason: system_limit
  in function  hipe_icode_range:inf_mult/2 (hipe_icode_range.erl, line 1627)
  in call from hipe_icode_range:range_mult/2 (hipe_icode_range.erl, line 1251)
  in call from hipe_icode_range:analyse_call_or_enter_fun/4 (hipe_icode_range.erl, line 1078)
  in call from hipe_icode_range:analyse_call/2 (hipe_icode_range.erl, line 410)
  in call from hipe_icode_range:analyse_insn/3 (hipe_icode_range.erl, line 251)
  in call from hipe_icode_range:analyse_BB/5 (hipe_icode_range.erl, line 241)
  in call from hipe_icode_range:analyse_block/4 (hipe_icode_range.erl, line 228)
  in call from hipe_icode_range:analyse_blocks/2 (hipe_icode_range.erl, line 214)
** exception exit: system_limit
     in function  hipe:run_compiler_1/4 (hipe.erl, line 695)
     in call from hipe:run_compiler/4 (hipe.erl, line 638)
     in call from hipe:c/3 (hipe.erl, line 332)
</pre>

Of course, this needs to be fixed, but it's most likely not related to exceptions.

With the following command, the compilation finishes, so the test can be checked for its exception behavior:
<pre>
6> c(exception_SUITE, [native, {hipe, [no_icode_range]}]).
exception_SUITE.erl:80: Warning: no clause will ever match
exception_SUITE.erl:81: Warning: the guard for this clause evaluates to 'false'
exception_SUITE.erl:84: Warning: this expression will fail with a 'badarith' exception
{ok,exception_SUITE}
</pre>

**Comment by bjorng** [comment]
I have re-run trycatch_SUITE. The `eclectic/1` test fails in the same place as it does without this PR. When running exception_SUITE compiled with to native code with the `no_icode_range` option, 6 test cases out of 11 fails. That is the same number as without this PR. I have not inspected the exact reasons for the failures, but given the nature of exception_SUITE, I assume that it fails because the differences of stack backtraces between native code and BEAM code.

So it seems that unless our daily builds uncover any problems, I can merge this PR soon.

**Comment by bjorng** [comment]
> I would generally say relying on the stacktrace entries for any reason other than debugging is a no-no, except maybe in a handful of cases such as the first entry of error:undef. Therefore I propose it is acceptable for native code/jit/erjang to provide less information in certain cases and, if so, we document those guarantees.

I agree. I think that has been the informal policy all along. It would be good to formally document it.

**Comment by josevalim** [comment]
> I agree. I think that has been the informal policy all along. It would be good to formally document it.

I will send a pull request to improve the docs of `erlang:get_stacktrace/3`. We can continue this separate discussion there.

**Pull request merged**